### ISSUE TITLE
Possible race condition between the Linkerd CNI and the AWS VPC CNI

### ISSUE THREAD

**Issue by JonKusz** [linkerd/linkerd2#10738]
### What is the issue?

When using the Linkerd CNI plugin in an AWS EKS cluster alongside the AWS VPC CNI, I notice the following behaviour when adding/replacing nodes.

On node startup, only the `linkerd-cni`, `aws-node`, and `kube-proxy` pods successfully start. All other pods fail to start and remain in either `Init` or `ContianerCreating` status, with the following error.

```
Warning  FailedCreatePodSandBox  2s (x2 over 14s)  kubelet            (combined from similar events): Failed to create pod sandbox: rpc error: code = Unknown desc = failed to setup network for sandbox "f9610b15286ede30cf3b88debb4ac0ee0a515555382ced008fac22e05562be77": plugin type="linkerd-cni" name="linkerd-cni" failed (add): cannot convert: no valid IP addresses
```

Looking at the logs for the `linkerd-cni` pod scheduled to the node, I see the following.

```
linkerd-cni/linkerd-cni-wbbx4[install-cni]: Wrote linkerd CNI binaries to /host/opt/cni/bin
linkerd-cni/linkerd-cni-wbbx4[install-cni]: No active CNI configuration files found; installing in "interface" mode in /host/etc/cni/net.d/01-linkerd-cni.conf
linkerd-cni/linkerd-cni-wbbx4[install-cni]: Using CNI config template from CNI_NETWORK_CONFIG environment variable.
linkerd-cni/linkerd-cni-wbbx4[install-cni]:       "k8s_api_root": "https://__kubernetes_service_host__/:__KUBERNETES_SERVICE_PORT__",
linkerd-cni/linkerd-cni-wbbx4[install-cni]:       "k8s_api_root": "https://172.20.0.1/:__KUBERNETES_SERVICE_PORT__",
linkerd-cni/linkerd-cni-wbbx4[install-cni]: CNI config: {
linkerd-cni/linkerd-cni-wbbx4[install-cni]:   "name": "linkerd-cni",
linkerd-cni/linkerd-cni-wbbx4[install-cni]:   "type": "linkerd-cni",
linkerd-cni/linkerd-cni-wbbx4[install-cni]:   "log_level": "info",
linkerd-cni/linkerd-cni-wbbx4[install-cni]:   "policy": {
linkerd-cni/linkerd-cni-wbbx4[install-cni]:       "type": "k8s",
linkerd-cni/linkerd-cni-wbbx4[install-cni]:       "k8s_api_root": "[https://172.20.0.1:443](https://172.20.0.1/)",
linkerd-cni/linkerd-cni-wbbx4[install-cni]:       "k8s_auth_token": "__SERVICEACCOUNT_TOKEN__"
linkerd-cni/linkerd-cni-wbbx4[install-cni]:   },
linkerd-cni/linkerd-cni-wbbx4[install-cni]:   "kubernetes": {
linkerd-cni/linkerd-cni-wbbx4[install-cni]:       "kubeconfig": "/etc/cni/net.d/ZZZ-linkerd-cni-kubeconfig"
linkerd-cni/linkerd-cni-wbbx4[install-cni]:   },
linkerd-cni/linkerd-cni-wbbx4[install-cni]:   "linkerd": {
linkerd-cni/linkerd-cni-wbbx4[install-cni]:     "incoming-proxy-port": 4143,
linkerd-cni/linkerd-cni-wbbx4[install-cni]:     "outgoing-proxy-port": 4140,
linkerd-cni/linkerd-cni-wbbx4[install-cni]:     "proxy-uid": 2102,
linkerd-cni/linkerd-cni-wbbx4[install-cni]:     "ports-to-redirect": [],
linkerd-cni/linkerd-cni-wbbx4[install-cni]:     "inbound-ports-to-ignore": ["4191","4190"],
linkerd-cni/linkerd-cni-wbbx4[install-cni]:     "simulate": false,
linkerd-cni/linkerd-cni-wbbx4[install-cni]:     "use-wait-flag": true
linkerd-cni/linkerd-cni-wbbx4[install-cni]:   }
linkerd-cni/linkerd-cni-wbbx4[install-cni]: }
linkerd-cni/linkerd-cni-wbbx4[install-cni]: Created CNI config /host/etc/cni/net.d/01-linkerd-cni.conf
linkerd-cni/linkerd-cni-wbbx4[install-cni]: Setting up watches.
linkerd-cni/linkerd-cni-wbbx4[install-cni]: Watches established.
```

Note that the CNI is starting in `interface` mode, which I interpret that to mean that the Linkerd CNI is not detecting the presence of the AWS VPC CNI.

After deleting the `linkerd-cni` pod scheduled to that node, I see the following logs in the new pod.

```
linkerd-cni/linkerd-cni-bckk4[install-cni]: Wrote linkerd CNI binaries to /host/opt/cni/bin
linkerd-cni/linkerd-cni-bckk4[install-cni]: Installing CNI configuration in "chained" mode for /host/etc/cni/net.d/10-aws.conflist
linkerd-cni/linkerd-cni-bckk4[install-cni]: Using CNI config template from CNI_NETWORK_CONFIG environment variable.
linkerd-cni/linkerd-cni-bckk4[install-cni]:       "k8s_api_root": "https://__kubernetes_service_host__/:__KUBERNETES_SERVICE_PORT__",
linkerd-cni/linkerd-cni-bckk4[install-cni]:       "k8s_api_root": "https://172.20.0.1/:__KUBERNETES_SERVICE_PORT__",
linkerd-cni/linkerd-cni-bckk4[install-cni]: CNI config: {
linkerd-cni/linkerd-cni-bckk4[install-cni]:   "name": "linkerd-cni",
linkerd-cni/linkerd-cni-bckk4[install-cni]:   "type": "linkerd-cni",
linkerd-cni/linkerd-cni-bckk4[install-cni]:   "log_level": "info",
linkerd-cni/linkerd-cni-bckk4[install-cni]:   "policy": {
linkerd-cni/linkerd-cni-bckk4[install-cni]:       "type": "k8s",
linkerd-cni/linkerd-cni-bckk4[install-cni]:       "k8s_api_root": "[https://172.20.0.1:443](https://172.20.0.1/)",
linkerd-cni/linkerd-cni-bckk4[install-cni]:       "k8s_auth_token": "__SERVICEACCOUNT_TOKEN__"
linkerd-cni/linkerd-cni-bckk4[install-cni]:   },
linkerd-cni/linkerd-cni-bckk4[install-cni]:   "kubernetes": {
linkerd-cni/linkerd-cni-bckk4[install-cni]:       "kubeconfig": "/etc/cni/net.d/ZZZ-linkerd-cni-kubeconfig"
linkerd-cni/linkerd-cni-bckk4[install-cni]:   },
linkerd-cni/linkerd-cni-bckk4[install-cni]:   "linkerd": {
linkerd-cni/linkerd-cni-bckk4[install-cni]:     "incoming-proxy-port": 4143,
linkerd-cni/linkerd-cni-bckk4[install-cni]:     "outgoing-proxy-port": 4140,
linkerd-cni/linkerd-cni-bckk4[install-cni]:     "proxy-uid": 2102,
linkerd-cni/linkerd-cni-bckk4[install-cni]:     "ports-to-redirect": [],
linkerd-cni/linkerd-cni-bckk4[install-cni]:     "inbound-ports-to-ignore": ["4191","4190"],
linkerd-cni/linkerd-cni-bckk4[install-cni]:     "simulate": false,
linkerd-cni/linkerd-cni-bckk4[install-cni]:     "use-wait-flag": true
linkerd-cni/linkerd-cni-bckk4[install-cni]:   }
linkerd-cni/linkerd-cni-bckk4[install-cni]: }
linkerd-cni/linkerd-cni-bckk4[install-cni]: Created CNI config /host/etc/cni/net.d/10-aws.conflist
linkerd-cni/linkerd-cni-bckk4[install-cni]: Setting up watches.
linkerd-cni/linkerd-cni-bckk4[install-cni]: Watches established.
```

Note that now the CNI is starting in `chained` mode, as I would have initially expected, and appending to the existing `10-aws.conflist` list that the VPC CNI creates.

Afterwards, the pods that were stuck in a bad state, either move to `Running` status, or do so after being recreated, as expected.

My hunch here is that there is some sort of race condition between the Linkerd CNI and the AWS VPC CNI where the Linkerd CNI starts before the AWS VPC CNI, leading to this issue. I've been able to reproduce this on the latest `2.13.0` release, which I understand includes some race condition mitigation, but seemingly only for meshed pods, whereas I believe this issue sits at the node level.

### How can it be reproduced?

Using the Helm installation method.

1. Add repo.
```
helm repo add linkerd https://helm.linkerd.io/stable
```

2. Install CRDs chart.
```
helm install linkerd-crds linkerd/linkerd-crds \
  -n linkerd --create-namespace
```

3. Install CNI plugin chart.
```
helm install linkerd-cni -n linkerd-cni --create-namespace --set useWaitFlag=true linkerd/linkerd2-cni
```

4. Install control plane chart.
```
helm install linkerd-control-plane \
  -n linkerd \
  --set-file identityTrustAnchorsPEM=ca.crt \
  --set-file identity.issuer.tls.crtPEM=issuer.crt \
  --set-file identity.issuer.tls.keyPEM=issuer.key \
  --set cniEnabled=true \
  linkerd/linkerd-control-plane
```

5. Replace a node.

### Logs, error output, etc

See logs included in the issue description.

### output of `linkerd check -o short`

```
linkerd-identity
----------------
‼ issuer cert is valid for at least 60 days
  issuer certificate will expire on 2023-04-16T02:19:15Z
  see https://linkerd.io/2.13/checks/#l5d-identity-issuer-cert-not-expiring-soon for hints

linkerd-webhooks-and-apisvc-tls
-------------------------------
‼ proxy-injector cert is valid for at least 60 days
  certificate will expire on 2023-04-13T12:19:15Z
  see https://linkerd.io/2.13/checks/#l5d-proxy-injector-webhook-cert-not-expiring-soon for hints
‼ sp-validator cert is valid for at least 60 days
  certificate will expire on 2023-04-13T12:19:14Z
  see https://linkerd.io/2.13/checks/#l5d-sp-validator-webhook-cert-not-expiring-soon for hints
‼ policy-validator cert is valid for at least 60 days
  certificate will expire on 2023-04-13T12:19:15Z
  see https://linkerd.io/2.13/checks/#l5d-policy-validator-webhook-cert-not-expiring-soon for hints

Status check results are √
```

### Environment

- Kubernetes Version: `1.24`
- Cluster Environment: `EKS`
- Linkerd Version: `stable-2.13.0`

### Would you like to work on fixing this bug?

no

**Comment by mateiidavid** [comment]
Can you confirm if the AWS CNI plugin version is `>= v1.12.6-eskbuild.1`, please?

**Comment by artem-collectai** [comment]
I experienced the same issue with 1.12.2

**Comment by JonKusz** [comment]
> Can you confirm if the AWS CNI plugin version is `>= v1.12.6-eskbuild.1`, please?

@mateiidavid I am currently using version `v1.12.1-eksbuild.2` of the VPC CNI. Would you suggest upgrading to `>= v1.12.6-eskbuild.1`?

**Comment by migueleliasweb** [comment]
I was also having this problem on version `v1.12.5-eksbuild.2` of the VPC CNI and EKS version 1.23.

**Comment by alpeb** [comment]
> I was also having this problem on version `v1.12.5-eksbuild.2` of the VPC CNI and EKS version 1.23.

It should be fixed in linkerd stable-2.12.5 and stable-2.13.2. Did you give those versions a try?

**Comment by migueleliasweb** [comment]
> > I was also having this problem on version `v1.12.5-eksbuild.2` of the VPC CNI and EKS version 1.23.
> 
> It should be fixed in linkerd stable-2.12.5 and stable-2.13.2. Did you give those versions a try?

My team recently upgraded the control plane, CRD and CNI helm charts to use the latest 2.13.2 release. Things are much more stable now. We've tested by doing some node rollouts in succession and we've found no more issues. 🚀

**Comment by alpeb** [comment]
@migueleliasweb great to hear that! Thanks for reporting back! :+1:
